src/python/pants/bin/daemon_pants_runner.py

```diff
@@ -12,36 +12,50 @@
 from builtins import open, zip
 from contextlib import contextmanager
 
-from future.utils import PY3, raise_with_traceback
-from setproctitle import setproctitle as set_process_title
-
 from pants.base.build_environment import get_buildroot
 from pants.base.exception_sink import ExceptionSink, SignalHandler
 from pants.base.exiter import PANTS_FAILED_EXIT_CODE, PANTS_SUCCEEDED_EXIT_CODE, Exiter
 from pants.bin.local_pants_runner import LocalPantsRunner
+from pants.init.logging import encapsulated_global_logger
 from pants.init.util import clean_global_runtime_state
 from pants.java.nailgun_io import NailgunStreamStdinReader, NailgunStreamWriter
-from pants.java.nailgun_protocol import ChunkType, NailgunProtocol
-from pants.pantsd.process_manager import ProcessManager
+from pants.java.nailgun_protocol import ChunkType, MaybeShutdownSocket, NailgunProtocol
 from pants.util.contextutil import hermetic_environment_as, stdio_as
 from pants.util.socket import teardown_socket
 
 
 class DaemonSignalHandler(SignalHandler):
 
   def handle_sigint(self, signum, _frame):
+    self.daemon.shutdown()
     raise KeyboardInterrupt('remote client sent control-c!')
 
+  def handle_sigterm(self, signum, _frame):
+    try:
+      self.daemon.shutdown()
+    except Exception:
+      pass
+
+
+class NoopExiter(Exiter):
+  def exit(self, result, *args, **kwargs):
+    if result != 0:
+      # TODO this exception was not designed for this, but it happens to do what we want.
+      raise _GracefulTerminationException(result)
+
 
 class DaemonExiter(Exiter):
-  """An Exiter that emits unhandled tracebacks and exit codes via the Nailgun protocol."""
+  """An Exiter that emits unhandled tracebacks and exit codes via the Nailgun protocol.
+
+  TODO: This no longer really follows the Exiter API, per-se (or at least, it doesn't call super).
```

Comment:
If this no longer follows the `Exiter` API, and (for now) we don't need the `DaemonExiter` to act as an exiter, we could maybe inline this code, or turn it into a standalone class and make this not inherit `Exiter` (possibly changing the name to `DaemonNailgunCloser` or something to avoid confusing this with other `*Exiter` classes.

Reply:
added todo in https://github.com/pantsbuild/pants/pull/7606